### ISSUE TITLE
What is the right way to use coverage.py with Tensorflow?

### ISSUE THREAD

**Issue by nedbat** [tensorflow/tensorflow#33759]
I apologize if this is the wrong way to ask this question. I'm the maintainer of coverage.py, for measuring code coverage in Python projects.  A user wrote an issue for me: https://github.com/nedbat/coveragepy/issues/856

After digging into it, I see that his tf.keras.Model.call() function is not executed directly, but is transformed into a temporary file, and executed there.  So coverage.py reports that his code is unexecuted, even though he can see the effects of its execution.

I also see that the transformed code has an `ag_source_map__` parameter which can be used to map back from the transformed code to the original code.  A coverage.py plugin could use that information to report coverage usefully.

My questions are:
1. Is there a reason people haven't reported this to coverage.py before? Is there a existing known way to get coverage reports on this kind of code?
2. What is a stable public API for getting the transformation mapping?
3. Would TensorFlow be interested in maintaining a coverage.py plugin to make this work properly?

**Comment by tmcclintock** [comment]
Has there been any progress on this on the TF side? This is an important feature for TF in production.

**Comment by csachs** [comment]
I second that request. This issue needs discussion and maybe special addressing on the ` coverage.py` side. In particular for TF2, all `tf.function` decorated functions are not observed by `coverage.py` (as a work-around, in some cases they can be dynamically un-decorated for unittesting)

**Comment by nedbat** [comment]
Does anyone want to take on writing a coverage.py plugin? I can help with the coverage.py side of things.

**Comment by csachs** [comment]
The mentioned issue ( https://github.com/Sujit-O/pykg2vec/issues/123 ) hinted me at the function [`tf.config.experimental_run_functions_eagerly(run_eagerly)`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/config/experimental_run_functions_eagerly), which looks like the ideal solution. However it does not seem to work completely as advertised, cf. https://github.com/tensorflow/tensorflow/issues/38170 , as certain functions cease working if switched between eager/non-eager modes.

**Comment by Zahlii** [comment]
With tensorflow 2.5, even if I run tf.config.experimental_run_functions_eagerly(True) directly before my tests, the coverage is not correctly reported. I want to write tests for a custom loss function around CRFs for a production model, where having adequate test coverage is key. Currently, I am stuck on having to exclude this file from coverage.py.

**Comment by nedbat** [comment]
I'm the coverage.py maintainer.  I'd be glad to work closely with someone from the Tensorflow side to find a solution to this problem.

**Comment by mohantym** [comment]
Hi @nedbat, Found some articles of unit testing and code coverage of Tensorflow codes , It indicates use of **tf.test api** .Attaching below for reference .
https://theaisummer.com/unit-test-deep-learning/
https://www.tensorflow.org/api_docs/python/tf/test/Benchmark
https://github.com/tensorflow/docs/blob/master/site/en/community/contribute/tests.md

**Comment by gsakkis** [comment]
Bump; this is still an issue.

**Comment by mohantym** [comment]
Ok! Reopening as requested.

**Comment by rivershah** [comment]
@fchollet  Any further progress or thoughts on this please? Coverage reports not being accurate is a production concern. Essentially any decorated function is showing up as a miss. Could we please look into prioritising this work with @gsakkis

**Comment by hmc-cs-mdrissi** [comment]
I think easiest fix is to just have a global way to fully disable autograph. `tf.config.experimental_run_functions_eagerly` doesn't actually disable all autograph usages. A layer called in v1 mode ignores that setting entirely. tf.data also looks to ignore that setting. Some way for tests under coverage to just fully opt out of autograph, tf.config.disable_autograph would be nice.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by hmc-cs-mdrissi** [comment]
The underlying issue hasn't been affected. The core way autograph works to create new source files in tensorflow currently does not interact well with coverage. This either requires,

1. A way to fully disable autograph. The existing apis that claim to don't do it fully.
2. A way to map back code to original source + coverage plugin.

The key relevant logic for 1 is [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/impl/api.py) and has not changed much in a year.

**Comment by Venkat6871** [comment]
Hi @nedbat  ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!

**Comment by nedbat** [comment]
@Venkat6871: you've commented twice on this issue. Both times, the text seemed like a form letter, which is fine, I'm sure Tensorflow gets a lot of issue traffic. But you have no picture, no description on your page, etc. It's hard for me to know what your role is here. I don't know how official your comments are, whether they represent the project as a whole or not. I can see that you have the ability to update the labels, so you've been granted permissions by the project, but it would still be better as a community-facing voice of the project to have a presence with more information.

**Comment by Venkat6871** [comment]
Hi @nedbat,
Thanks for the feedback and for pointing this out. You are right, this thread deserved a more careful/manual response from my side, and I understand your concern regarding the repeated template-style replies.
I will update the labels accordingly. Also appreciate your continued input around the `coverage.py` integration discussion.

**Comment by mihaimaruseac** [comment]
Venkat is not an LLM. Is a contractor. The real OSS team is not handling GitHub anymore.

**Comment by nedbat** [comment]
Just to be clear, I never thought or claimed that Venkat was an LLM. I'm just trying to understand the group dynamics here, and make it more apparent to people coming through this repo infrequently.

**Comment by mihaimaruseac** [comment]
Yeah, so since 2022 the remaining TF OSS team started moving away from the repo. Delegating more and more responsibilities to contractors, "forgetting" to update issues, reducing the number of patch releases and now full releases. Just look at #102890 as the most recent example.
